networking_cisco/plugins/cisco/db/l3/l3_router_appliance_gbp_db.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2014 Cisco Systems, Inc.  All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from networking_cisco.plugins.cisco.db.l3 import (
+    l3_router_appliance_db as l3_r_db)
+from networking_cisco.plugins.cisco.l3.drivers import driver_context
+
+from neutron.common import constants as l3_constants
+from neutron.common import exceptions as n_exc
+from neutron.i18n import _LW
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+FLOATINGIP_STATUS_ACTIVE = l3_constants.FLOATINGIP_STATUS_ACTIVE
+
+LOG = logging.getLogger(__name__)
+
+
+class L3RouterApplianceGbpDBMixin(l3_r_db.L3RouterApplianceDBMixin):
+    """Mixin class implementing Neutron's routing service using appliances.
+
+    This class overrides the base superclass, and should only be used for
+    integrations with ACI using the Group Based Policy workflow.
+    """
+
+    def create_floatingip(self, context, floatingip,
+                          initial_status=FLOATINGIP_STATUS_ACTIVE):
+        """Override of parent class's create_floatingip.
+
+        This override of the parent class's create_floatingip assumes
+        is needed for GBP workflow, as the pre-/post-commmit calls for
+        creating the floating IP must be peformed in a loop with the database
+        call in the parent class.
+        """
+
+        res = None
+        fip = floatingip['floatingip']
+        if not fip.get('subnet_id'):
+            # NOTE: default router type must be ASR1k
+            router_type_name = cfg.CONF.routing.default_router_type
+            driver = self._get_router_type_driver(context,
+                                                  router_type_name)
+            if driver:
+                fip_ctxt = driver_context.FloatingipContext(floatingip)
+                driver.create_floatingip_precommit(context, fip_ctxt)
+                nat_pool_list = getattr(context, 'nat_pool_list', [])
+                for nat_pool in nat_pool_list:
+                    if not nat_pool:
```

Comment:
I think I found in my UT that None can appear in the list, which is why this is there. For example:
>>> mylist = ['1', '2', None, '3']
>>> for element in mylist:
...     print element
... 
1
2
None
3